Bluetooth: Host: Fix re-pairing across local ids

alwa-nordic opened the pull request:
This PR prevents multiple bonds to the same peer with different local identities. This is not possible (at least without some complicated workarounds) due to limitations in the resolve list. Before this change, the host would accept multiple bonds, but the controller would fail to add all bonds to the resolve list. This inconsistency caused multiple issues.

This PR adds a new config "BT_ID_UNPAIR_MATCHING_BONDS" to give a behavior similar to re-bonding in this situation. When a bond is about to complete, this config will make the stack find any other bond with the same peer address (or IRK) and `bt_unpair` that bond before the event `pairing_complete`.

This PR adds Babblesim tests.

See the commit messages for more detail.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22049

alwa-nordic committed:
Bluetooth: Host: Add `CONFIG_BT_ID_UNPAIR_MATCHING_BONDS`

This commit prevents the assertion error added in the previous commit,
by introducing the following behavior.

When the Host detects that a pairing procedue would result in more than
one local identity having a bond with the same remote address, it will
try to abort the pairing. If the pairing procedure cannot be properly
aborted, it will remove the pairing locally and disconnect.

This commit also introduces a new kconfig `BT_ID_UNPAIR_MATCHING_BONDS`.
This config changes the above-described behavior to instead un-pair the
old bond and continue with the pairing (if the new bond has the same or
better security.)

The new kconfig is not enabled by default. See the help text of the new
kconfig option for more details.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/id.c:
```diff
@@ -788,13 +793,62 @@ void bt_id_pending_keys_update(void)
+	struct bt_keys *found;
+};
+
+/* The Controller Resolve List is constrained by 7.8.38 "LE Add Device To Resolving List command".
```
reflow to 80col

jori-nordic commented on subsys/bluetooth/host/id.c:
```diff
@@ -788,13 +793,62 @@ void bt_id_pending_keys_update(void)
 	BT_DBG("addr %s", bt_addr_le_str(&keys->addr));
 
+	__ASSERT_NO_MSG(keys != NULL);
+	__ASSERT_NO_MSG(!bt_id_find_conflict(keys).found);
```
an assert for this seems pretty harsh. I would rather return an error code.

alwa-nordic replied:
I'm just documenting the assumption. It has always been there. Breaking this assumption was the source of this bug.

I did contemplate changing this function to "try add". But I don't think it's prettier. I suppose we would call it like in the following example?
```c
int err;
do {
    err = bt_id_add(new_bond);
    if (err == -EADDRINUSE) {
        bool allowed;
        int unpair_err;
	struct bt_keys conflict;

        BT_INFO("New bond conflict.");
	conflict = bt_id_find_conflict(new_bond);
        __ASSERT_NO_MSG(conflict);
        allowed = update_keys_check(smp, conflict.found);
        if (!allowed) {
            BT_WARN("Bond overwrite denied. Refusing new pairing.");
            return BT_SMP_ERR_AUTH_REQUIREMENTS;
        }
        BT_INFO("Bond overwrite permitted. Un-pairing old conflicting bond and finalizing new.");
        unpair_err = bt_unpair(conflict->id, &conflict->addr);
        __ASSERT_NO_MSG(!unpair_err);
    }
} while(err == -EADDRINUSE);
```

jori-nordic commented on subsys/bluetooth/host/keys.c:
```diff
@@ -6,6 +6,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include "zephyr/sys/__assert.h"
```
<>

jori-nordic commented on subsys/bluetooth/host/keys.h:
```diff
@@ -6,6 +6,9 @@
  */
 
+#include "zephyr/bluetooth/addr.h"
+#include "zephyr/sys/util_macro.h"
```
<>

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3928,6 +3930,38 @@ static uint8_t smp_ident_info(struct bt_smp *smp, struct net_buf *buf)
 	return 0;
 }
 
+/* For now, the SMP module makes the decision on behalf of the user to any
```
missing `replace`, `overwrite` or `remove` ?

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3995,7 +4029,10 @@ static uint8_t smp_ident_addr_info(struct bt_smp *smp, struct net_buf *buf)
 		}
 
-		bt_id_add(keys);
+		err = smp_id_add_unpair(smp, keys);
```
should you also change https://github.com/zephyrproject-rtos/zephyr/blob/a4c046cd5a89a513483d164a17ec0e3178be07dc/subsys/bluetooth/host/smp.c#L4786 ?

alwa-nordic replied:
I think yes. At least it can't hurt. I'll see if I can make a bablesim test that fails there.

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3928,6 +3930,38 @@ static uint8_t smp_ident_info(struct bt_smp *smp, struct net_buf *buf)
+ *
+ * This logic should be moved out to the application. But it requires new API.
+ */
+static uint8_t smp_id_add_unpair(struct bt_smp *smp, struct bt_keys *new_bond)
```
I'd name that one `smp_id_add_replace`

jori-nordic replied:
bump

alwa-nordic committed:
Bluetooth: Host: Define `bt_irk_eq`

This gives a well-defined method for determining if two IRKs are equal.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Prevent bonding to the same device more than once

This change prevents two local identities from having bonds to the same
device.

The Core specification is not well suited for Zephyr's multiple-local-
identities feature. The HCI specification seems written with intent that
a controller is used for only one GAP device. A GAP device has at most
one public address, and at most one random static address.

The Zephyr Bluetooth API, on the other hand, has a concept of local
identities. This feature allows the Zephyr Bluetooth stack to
simulatainously assume multiple local addresses. This does not mesh well
with the above intent in the specification.

In particular, the HCI specification for the resolve list does not allow
more than one entry for a remote address. The controller will deny any
attempts at doing this.

The current implementation of the Zephyr host will try the above and be
denied. But there is no handling for this situation and the host ends up
in a confused state. Some parts of the system are ok with the two bonds,
but other parts assume this situation never occurs behave badly.

The result is that the host confuses the multiple bonds to the same
device. Symtoms include:

- Directed advertisements have a different source address than what the
  host intended, in which case the two sides are confused about the
  address of the Zephyr advertiser, and as a result LTKs will not match.

- Errors in the log.

This commit simply asserts. This is not a solution, just a placeholder
for a fix. The next commit will implement a strategy for handling this
situation instead of failing this assert.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Host: BS tests: bsim_test_bond_overwrite_allowed/denied

These tests prove the existence of a bug in the local identity feature.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -466,6 +466,43 @@ config BT_SMP_ALLOW_UNAUTH_OVERWRITE
 	  to create a new bond the old bond has to be explicitly deleted with
 	  bt_unpair.
 
+config BT_ID_UNPAIR_MATCHING_BONDS
```
Should we s/MATCHING/CONFLICTING since the API you have below uses this language?

alwa-nordic replied:
As discussed oob, I'll keep it as is.

jori-nordic commented on subsys/bluetooth/host/id.c:
```diff
@@ -788,13 +792,70 @@ void bt_id_pending_keys_update(void)
 	BT_DBG("addr %s", bt_addr_le_str(&keys->addr));
 
+	__ASSERT_NO_MSG(keys != NULL);
+	__ASSERT_NO_MSG(!bt_id_find_conflict(keys));
```
I think i'd rather log an error and exit. Don't wanna have an assert failing in the wild because of something we missed.
We can clean up the bonding procedure later.

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3928,6 +3949,48 @@ static uint8_t smp_ident_info(struct bt_smp *smp, struct net_buf *buf)
+			return BT_SMP_ERR_AUTH_REQUIREMENTS;
+		}
+
+		BT_INFO("Bond replace permitted. Un-pairing old conflicting bond and finalizing new.");
```
I think you can leave out. "Bond replace permitted.".

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3928,6 +3949,48 @@ static uint8_t smp_ident_info(struct bt_smp *smp, struct net_buf *buf)
+		return BT_SMP_ERR_AUTH_REQUIREMENTS;
+	}
+
+	if (conflict && IS_ENABLED(BT_ID_UNPAIR_MATCHING_BONDS)) {
```
Are you sure you're exercising this branch in your test? I think it needs the `CONFIG_` prefix, I just used `IS_ENABLED` recently.

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -4739,14 +4805,34 @@ static void bt_smp_encrypt_change(struct bt_l2cap_chan *chan,
 		atomic_set_bit(smp->allowed_cmds, BT_SMP_CMD_SIGNING_INFO);
 	}
 
+	/* This is the latest point that is common for all code paths in the
```
last point

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -4739,14 +4805,34 @@ static void bt_smp_encrypt_change(struct bt_l2cap_chan *chan,
+	 * pairing process (during which we still have the option to send
+	 * Pairing Failed). That makes it convenient to update the RL here. We
+	 * want to update the RL during the pairing process so that we can fail
+	 * it in case there is a conflict with an already-existing bond.
```
s/already-//

jori-nordic reviewed (changes requested):
some comments. Im a bit concerned that you have tests but they're not failing, they should be since the `IS_ENABLED()` call should not work.

alwa-nordic commented:
> some comments. Im a bit concerned that you have tests but they're not failing, they should be since the `IS_ENABLED()` call should not work.

I fixed the IS_ENABLED. The test's were failing without the fix.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -1984,6 +1985,14 @@ static int smp_error(struct bt_smp *smp, uint8_t reason)
+	/* By spec, SMP "pairing process" completes successfully when the last
+	 * key to distribute is acknowledged at link-layer.
+	 */
+	remote_already_completed = (atomic_test_bit(smp->flags, SMP_FLAG_KEYS_DISTR)
+	                            && !smp->local_dist
+				    && !smp->remote_dist);
```
Minor style issue: the line splits should happen right after `&&`, not right before it

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3930,6 +3951,48 @@ static uint8_t smp_ident_info(struct bt_smp *smp, struct net_buf *buf)
+	conflict = bt_id_find_conflict(new_bond);
+
+	if (conflict) {
+		BT_INFO("New bond conflicts with a bond on id %d.", conflict->id);
```
Seems like this and the other `BT_INFO()` in this function should instead be `BT_DBG()`?

jhedberg commented on include/zephyr/bluetooth/addr.h:
```diff
@@ -81,6 +81,14 @@ static inline int bt_addr_cmp(const bt_addr_t *a, const bt_addr_t *b)
+static inline bool bt_addr_le_eq(const bt_addr_le_t *a, const bt_addr_le_t *b)
+{
+	return ((a->type == b->type)
+	        && (memcmp(a->a.val, b->a.val, sizeof(a->a.val)) == 0));
```
Same thing here with the placement of `&&`

alwa-nordic replied:
This will disappear when I rebase on https://github.com/zephyrproject-rtos/zephyr/pull/50361.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3930,6 +3951,48 @@ static uint8_t smp_ident_info(struct bt_smp *smp, struct net_buf *buf)
+	__ASSERT_NO_MSG(!(smp->remote_dist & BT_SMP_DIST_ID_KEY));
+
+	conflict = bt_id_find_conflict(new_bond);
+
```
You can remove this line (generally: no empty line between assignment to a variable and testing for its value)

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3930,6 +3951,48 @@ static uint8_t smp_ident_info(struct bt_smp *smp, struct net_buf *buf)
+		int unpair_err;
+
+		trust_ok = update_keys_check(smp, conflict);
+
```
style: remote the empty line
